hooks/nvidia-bootstrap/image/run.sh

```diff
@@ -12,50 +12,219 @@
+  chroot ${ROOTFS_DIR} /bin/bash -c 'apt-get --no-upgrade -y install gcc linux-headers-$(uname -r)'
+fi
+
+if [[ ! -f ${ROOTFS_DIR}/usr/bin/gcc-7 ]]; then
+  echo "Installing gcc-7 on host machine"
+
+  # Temporarily add the debian "buster" repo where gcc-7 lives
+  #   But first clear it out first if it already exists
+  sed -n '/buster/q;p' -i ${ROOTFS_DIR}/etc/apt/sources.list
+  echo "deb http://deb.debian.org/debian buster main" >> ${ROOTFS_DIR}/etc/apt/sources.list
+
+  # Install gcc-7
+  chroot ${ROOTFS_DIR} apt-get update
+  chroot ${ROOTFS_DIR} /bin/bash -c 'apt-get -y install linux-headers-$(uname -r)'
+  chroot ${ROOTFS_DIR} /bin/bash -c 'DEBIAN_FRONTEND=noninteractive apt-get -t buster --no-upgrade -y install gcc-7'
+
+  # Remove the debian "buster" repo line that was added above
+  sed -n '/buster/q;p' -i ${ROOTFS_DIR}/etc/apt/sources.list
+  chroot ${ROOTFS_DIR} apt-get update
+fi
+
+# Unload open-source nouveau driver if it exists
+#   The nvidia drivers won't install otherwise
+#   "g3" instances in particular have this module auto-loaded
+chroot ${ROOTFS_DIR} modprobe -r nouveau || true
+
+
+#################################################
+# Download and install the Nvidia drivers and cuda libraries
 
-  chroot /rootfs mknod -m 666 /dev/nvidia-uvm c $D 0
-else
+# Create list of URLs and Checksums by merging driver item with array of cuda files
+downloads=(${class_to_driver_file[$AWS_INSTANCE_CLASS]} ${cuda_files[@]})
+checksums=(${class_to_driver_checksum[$AWS_INSTANCE_CLASS]} ${cuda_files_checksums[@]})
+
+# Ensure that the cache directory exists
+mkdir -p $CACHE_DIR_CONTAINER
+
+# Download, verify, and execute each file
+length=${#downloads[@]}
+for (( i=0; i<${length}; i++ )); do
+  download=${downloads[$i]}
+  checksum=${checksums[$i]}
+  filename=$(basename $download)
+  filepath_host="${CACHE_DIR_HOST}/${filename}"
+  filepath_container="${CACHE_DIR_CONTAINER}/${filename}"
+  filepath_installed="${CACHE_DIR_CONTAINER}/${filename}.installed"
+
+  echo "Checking for file at $filepath_container"
+  if [[ ! -f $filepath_container ]] || ! (echo "$checksum  $filepath_container" | sha1sum -c - 2>&1 >/dev/null); then
+    echo "Downloading $download"
+    curl -L $download > $filepath_container
+    chmod a+x $filepath_container
+  fi
+
+  echo "Verifying sha1sum of file at $filepath_container"
+  if ! (echo "$checksum  $filepath_container" | sha1sum -c -); then
+    echo "Failed to verify sha1sum for file at $filepath_container"
+    exit 1
+  fi
+
+  # Install the Nvidia driver and cuda libs
+  if [[ -f $filepath_installed ]]; then
+    echo "Detected prior install of file $filename on host"
+  else
+    echo "Installing file $filename on host"
+    if [[ $download =~ .*NVIDIA.* ]]; then
+      # Install the nvidia package (using gcc-7)
+      chroot ${ROOTFS_DIR} /bin/bash -c "CC=/usr/bin/gcc-7 $filepath_host --accept-license --silent"
```

Comment:
Are these options (with the default directory) compatible with [`deviceplugins`](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/)

Reply:
Also for the deviceplugins this is the Google approach (alternative to the Nvidia one) [to prepare the node](https://github.com/GoogleCloudPlatform/container-engine-accelerators/pull/51)

Reply:
See also https://github.com/GoogleCloudPlatform/container-engine-accelerators/tree/master/cmd/nvidia_gpu

Reply:
This patch does not take the deviceplugin approach, but it also does not preclude it.  It does not install nvidia-docker, and it does not swap the default container runtime.  Those could be added on top if we wanted to implement the deviceplugins over the continuation of the existing method.  That could be a pull request on top.

The linked [PR](https://github.com/GoogleCloudPlatform/container-engine-accelerators/pull/51) is interesting because it installs drivers via a daemonset.  If one didn't mind running containers on Kubernetes in privileged mode, it would be an interesting alternative to kops hooks.  It's also nice because one could deploy via helm chart rather than editing kops instancegroup manifests.

I did take a look at the Google setup [script](https://github.com/GoogleCloudPlatform/cos-gpu-installer/blob/master/entrypoint.sh), hoping to ditch what I just wrote in this PR.  Unfortunately, just like this current PR, the setup instructions are cloud specific.

Reply:
The issue is that accellerator is [already deprecated](https://github.com/kubernetes/kubernetes/pull/57384).

Reply:
https://github.com/kubernetes/kubernetes/pull/61498

Reply:
Yes for now just to care if there is some particular directoroes setup needed as a parameter by the Nvidia installer that could help deviceplugins support. If you check I think that the Google solution it is trying to pass good options to the Nvidia installer.